dask/array/routines.py

```diff
@@ -332,6 +333,52 @@ def vdot(a, b):
     return dot(a.conj().ravel(), b.ravel())
 
 
+def _shape_minus_axis(shape, axis):
+    newshape = list(shape)
+    newshape.pop(axis)
+    return tuple(newshape)
+
+
+def _chunk_sum(a, axis=None, dtype=None, keepdims=None):
+    # Caution: this is not your conventional array-sum: due
+    # to the special nature of the preceding blockwise con-
+    # traction,  each chunk is expected to have exactly the
+    # same shape,  with a size of 1 for the dimension given
+    # by `axis` (the reduction axis).  This makes mere ele-
+    # ment-wise addition of the arrays possible.   Besides,
+    # the output can be merely reshaped to lose the `axis`-
+    # dimension when keepdims = False
+    if type(a) is list:
+        xp = np
+
+        if is_cupy_type(a[0]):
+            import cupy
+
+            xp = cupy
+
+        out = reduce(partial(xp.add, dtype=dtype), a)
+    else:
+        if a.shape == (0,):
+            return a
```

Comment:
What is this case? And does it need to be a separate branch?

Reply:
This branch turns out to be necessary, as `_chunk_sum` is often called with an empty array `a` parameter passed in order to compute meta data &mdash; a peculiarity of `dask.array.reduction`

Reply:
@ParticularMiner if you remove that branch and let it go into the other branch, eventually into the `reduce`/`add`, what happens, how does it fail?

Reply:
I get an `IndexError: too many indices for array: array is 1-dimensional, but 2 were indexed` 
at the `reduce`/`add` line when computing the meta data.   Would you like the full traceback log?

Reply:
Here's the error log from one test:
[tracebacklog.txt](https://github.com/dask/dask/files/7645059/tracebacklog.txt)

Reply:
The problem is that the error occurs in dask *outside* any of the code that I wrote in this PR.

Reply:
@ParticularMiner thanks, oh, so this error happens only for a specific case: `[(7,), (7,), (), ()],`, which will reduce to a  scalar, are you sure there isn't something going on here? 🤔

Reply:
Oh no.  It's not the only case.  I just uploaded the error for one case. There are at least 15 other cases where the error occurs for that test unit `test_matmul()`.